src/core/mavlink_commands.cpp

```diff
@@ -318,5 +322,101 @@ void MAVLinkCommands::call_callback(
     // we lock ourselves out when we send a command in the callback receiving a command result.
     _parent.call_user_callback([callback, result, progress]() { callback(result, progress); });
 }
+// MavlinkCommandSender class
+
+// MavlinkCommandReceiver class
+MavlinkCommandReceiver::MavlinkCommandReceiver(SystemImpl& parent) : _parent(parent)
+{
+    _parent.register_mavlink_message_handler(
+        MAVLINK_MSG_ID_COMMAND_LONG,
+        std::bind(&MavlinkCommandReceiver::receive_command_long, this, std::placeholders::_1),
+        this);
+
+    _parent.register_mavlink_message_handler(
+        MAVLINK_MSG_ID_COMMAND_INT,
+        std::bind(&MavlinkCommandReceiver::receive_command_int, this, std::placeholders::_1),
+        this);
+}
+
+MavlinkCommandReceiver::~MavlinkCommandReceiver()
+{
+    unregister_all_mavlink_command_handlers(this);
+
+    _parent.unregister_all_mavlink_message_handlers(this);
+}
+
+void MavlinkCommandReceiver::receive_command_int(mavlink_message_t message)
+{
+    mavlink_command_int_t command_int;
+    mavlink_msg_command_int_decode(&message, &command_int);
+    MavlinkCommandReceiver::Command cmd(command_int);
+
+    std::lock_guard<std::mutex> lock(_mavlink_command_handler_table_mutex);
+
+    for (auto it = _mavlink_command_handler_table.begin();
+         it != _mavlink_command_handler_table.end();
+         ++it) {
+        if (it->cmd_id == command_int.command) {
+            it->callback(cmd);
+        }
+    }
+}
+
+void MavlinkCommandReceiver::receive_command_long(mavlink_message_t message)
```

Comment:
Usually we don't copy the mavlink_message_t's but use `const mavlink_message_t&`. I think that could be done here as well.